x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java

```diff
@@ -373,44 +386,105 @@ static void checkFullModelDefinitionIsPresent(
         final String modelId = config.getModelId();
 
         String index = ((IndexLocation) config.getLocation()).getIndexName();
-        mlOriginClient.prepareSearch(index)
-            .setQuery(
-                QueryBuilders.constantScoreQuery(
-                    QueryBuilders.boolQuery()
-                        .filter(QueryBuilders.termQuery(TrainedModelConfig.MODEL_ID.getPreferredName(), modelId))
-                        .filter(
-                            QueryBuilders.termQuery(InferenceIndexConstants.DOC_TYPE.getPreferredName(), TrainedModelDefinitionDoc.NAME)
-                        )
-                )
-            )
-            .setFetchSource(false)
-            .addDocValueField(TrainedModelDefinitionDoc.DEFINITION_LENGTH.getPreferredName())
-            .addDocValueField(TrainedModelDefinitionDoc.TOTAL_DEFINITION_LENGTH.getPreferredName())
-            .setSize(MAX_NUM_NATIVE_DEFINITION_PARTS)
-            .setTrackTotalHits(true)
-            .addSort(SortBuilders.fieldSort(TrainedModelDefinitionDoc.DOC_NUM.getPreferredName()).order(SortOrder.ASC).unmappedType("long"))
-            .execute(ActionListener.wrap(response -> {
-                SearchHit[] hits = response.getHits().getHits();
-                if (hits.length == 0) {
+
+        // Step 3
+        // Verify the model definition parts are all present
+        ActionListener<SearchResponse> step3SearchResultsVerificationListener = step3VerifyModelPartsArePresent(
```

Comment:
This would probably be a larger refactor but there's a new way to implement the chained action listeners. Here's an example: https://github.com/elastic/elasticsearch/pull/99623

Reply:
wow thats awesome! I will definitely do that next time